bin/install.sh

```
@@ -0,0 +1,198 @@
+    GREEN=$(printf '\033[32m')
+    YELLOW=$(printf '\033[33m')
+    BLUE=$(printf '\033[34m')
+    MAGENTA=$(printf '\033[35m')
+    BOLD=$(printf '\033[1m')
+    RESET=$(printf '\033[m')
+  else
+    RED=""
+    GREEN=""
+    YELLOW=""
+    BLUE=""
+    MAGENTA=""
+    BOLD=""
+    RESET=""
+  fi
+}
+
+get_os() {
+  case "$(uname -s)" in
+    *linux* ) echo "Linux" ;;
+    *Linux* ) echo "Linux" ;;
+    *darwin* ) echo "Darwin" ;;
+    *Darwin* ) echo "Darwin" ;;
+  esac
+}
+
+get_machine() {
+  case "$(uname -m)" in
+    "x86_64"|"amd64"|"x64")
+      echo "x86_64" ;;
+    "i386"|"i86pc"|"x86"|"i686")
+      echo "i386" ;;
+    "arm64"|"armv6l"|"aarch64")
+      echo "arm64"
+  esac
+}
+
+get_tmp_dir() {
+  echo $(mktemp -d)
+}
+
+do_checksum() {
+  checksum_url=$(get_checksum_url $version)
+  get_checksum_url $version
+  expected_checksum=$(curl -sL $checksum_url | grep $asset_name | awk '{print $1}')
+
+  if command_exists sha256sum; then
+    checksum=$(sha256sum $asset_name | awk '{print $1}')
+  elif command_exists shasum; then
+    checksum=$(shasum -a 256 $asset_name | awk '{print $1}')
+  else
+    fmt_warning "Could not find a checksum program. Install shasum or sha256sum to validate checksum."
+    return 0
+  fi
+
+  if [ "$checksum" != "$expected_checksum" ]; then
+    fmt_error "Checksums do not match"
+    exit 1
+  fi
+}
+
+do_install_binary() {
+  asset_name=$(get_asset_name $os $machine)
+  download_url=$(get_download_url $version $os $machine)
+
+  command_exists curl || {
+    fmt_error "curl is not installed"
+    exit 1
+  }
+
+  command_exists tar || {
+    fmt_error "tar is not installed"
+    exit 1
+  }
+
+  local tmp_dir=$(get_tmp_dir)
+
+  # Download tar.gz to tmp directory
+  echo "Downloading $download_url"
+  (cd $tmp_dir && curl -sL -O "$download_url")
+
+  (cd $tmp_dir && do_checksum)
+
+  # Extract download
+  (cd $tmp_dir && tar -xzf "$asset_name")
+
+  # Install binary
+  mv "$tmp_dir/$BINARY_NAME" $INSTALL_DIR
```

```suggestion
  # Create installation directory if it doesn't exist
  if [ ! -d "$INSTALL_DIR" ]; then
    echo "Creating installation directory $INSTALL_DIR"
    if ! mkdir -p "$INSTALL_DIR" 2>/dev/null; then
      echo "Permission denied: cannot create directory $INSTALL_DIR"
      echo "Attempting to create with sudo..."
      sudo mkdir -p "$INSTALL_DIR" || {
        fmt_error "Failed to create $INSTALL_DIR, even with sudo"
        exit 1
      }
    fi
  fi

  # Check if we have write permissions to the target directory
  if [ -w "$INSTALL_DIR" ]; then
    mv "$tmp_dir/$BINARY_NAME" "$INSTALL_DIR"
  else
    echo "Permission denied: cannot write to $INSTALL_DIR"
    echo "Attempting to install with sudo..."
    if command_exists sudo; then
      sudo mv "$tmp_dir/$BINARY_NAME" "$INSTALL_DIR" || {
        fmt_error "Failed to install to $INSTALL_DIR, even with sudo"
        echo "Try running this script with sudo or specifying a different directory:"
        echo "    INSTALL_DIR=\$HOME/bin ./bin/install.sh"
        exit 1
      }
    else
      fmt_error "Cannot write to $INSTALL_DIR and sudo is not available"
      echo "Try running this script as root or specifying a different directory:"
      echo "    INSTALL_DIR=\$HOME/bin ./bin/install.sh"
      exit 1
    fi
  fi

  # Make the binary executable
  if [ -w "$INSTALL_DIR/$BINARY_NAME" ]; then
    chmod +x "$INSTALL_DIR/$BINARY_NAME"
  else
    sudo chmod +x "$INSTALL_DIR/$BINARY_NAME" 2>/dev/null || {
      fmt_error "Could not make $INSTALL_DIR/$BINARY_NAME executable"
      exit 1
    }
  fi

```

---

Personally I would fail the script with a helpful message if you don't have write perms in the target dir.

---

Yeah, I agree. I'd rather fail the script if we're missing permissions and provide a new error message. I've pushed a change according.